### PR TITLE
netty: Requests with Connection header are malformed

### PR DESCRIPTION
**Pull request by ejona86**
Although this is part of HTTP/2 and should have already been handled
already, it was noticed as part of RBAC work to avoid matching
hop-by-hop headers. See gRFC A41.

CC @YifeiZhuang

### CODE REVIEW

**Review by dapengzhang0 — APPROVED** [review]
LGTM. However, there is a concern similar to the issues of Cloud Datastore emulator when upgrading Netty. Those libraries use a HttpAdapter that blindly (and mistakenly) copies metadata in HTTP1.1 request and convert it to gRPC client request.

- Old clients that worked for a decade suddenly stop to work when grpc-java server is upgraded to new version
  - We can hardly do anything about that other than a heads-up in release announcement.
- It's hard to debug when seeing the error on server side.
- Client does not work even if upgrading grpc.

We may do the following on client side
- Let grpc client throw if there is illegal headers in metadata. This helps debugging but the client still can not work even if upgrading grpc.
- Let grpc client (cross-language) discard illegal headers in metadata.

**Review by dapengzhang0 — APPROVED** [review]
Logging waring when creating the metada can help debug a lot. It would be even better if we do this cross-language. 

>> It's hard to debug when seeing the error on server side.

> There is no error on server-side.

Yeah, there is no error on server-side, I mean it's harder to debug when see the error when client receives error from the server rather than when sending the request.

File: api/src/main/java/io/grpc/Metadata.java [hunk 1]
```diff
@@ -733,6 +736,15 @@ private static BitSet generateValidTChars() {
     private static String validateName(String n, boolean pseudo) {
       checkNotNull(n, "name");
       checkArgument(!n.isEmpty(), "token must have at least 1 tchar");
+      if (n.equals("connection")) {
```
Comment (dapengzhang0) [inline api/src/main/java/io/grpc/Metadata.java:739]: Is this always lower case?

Comment (ejona86) [reply api/src/main/java/io/grpc/Metadata.java:739]: Yep. https://github.com/grpc/grpc-java/blob/7c6f53ab79846decde5e624bb6fa5f7a1f8a922c/api/src/main/java/io/grpc/Metadata.java#L750

**Pull request merged**